elasticapm/contrib/serverless/__init__.py

```diff
@@ -0,0 +1,193 @@
+    # transaction in __exit__/__enter__
+
+    def __init__(self, **kwargs):
+        self.name = kwargs.get("name")
+        self.event = {}
+        self.context = {}
+        self.response = None
+        self.client = ServerlessClient(**kwargs)
+        if not self.client.config.debug and self.client.config.instrument:
+            elasticapm.instrument()
+
+    def __call__(self, func):
+        self.name = self.name or get_name_from_func(func)
+
+        @functools.wraps(func)
+        def decorated(*args, **kwds):
+            if len(args) == 2:
+                # Saving these for request context later
+                self.event, self.context = args
+            else:
+                self.event, self.context = {}, {}
+            if not self.client.config.debug and self.client.config.instrument:
+                with self:
+                    self.response = func(*args, **kwds)
+                    return self.response
+            else:
+                return func(*args, **kwds)
+
+        return decorated
+
+    def __enter__(self):
+        """
+        Transaction setup
+        """
+        trace_parent = TraceParent.from_headers(self.event.get("headers"))
+        self.transaction = self.client.begin_transaction("request", trace_parent=trace_parent)
+        elasticapm.set_context(
+            lambda: get_data_from_request(
+                self.event,
+                capture_body=self.client.config.capture_body in ("transactions", "all"),
+                capture_headers=self.client.config.capture_headers,
+            ),
+            "request",
+        )
+        if "httpMethod" in self.event and "resource" in self.event:
+            elasticapm.set_transaction_name("{} {}".format(self.event["httpMethod"], self.event["resource"]))
+        else:
+            elasticapm.set_transaction_name(self.name, override=False)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Transaction teardown
+        """
+        if exc_val:
+            self.client.capture_exception(exc_info=(exc_type, exc_val, exc_tb), handled=False)
+
+        if self.response and isinstance(self.response, dict):
+            elasticapm.set_context(
+                lambda: get_data_from_response(self.response, capture_headers=self.client.config.capture_headers),
+                "response",
+            )
+            if "statusCode" in self.response:
+                result = "HTTP {}xx".format(int(self.response["statusCode"]) // 100)
+                elasticapm.set_transaction_result(result, override=False)
+        self.client.end_transaction()
+
+
+def get_data_from_request(event, capture_body=False, capture_headers=True):
+    """
+    Capture context data from API gateway event
+    """
+    result = {}
+    if capture_headers and "headers" in event:
+        result["headers"] = event["headers"]
+    if capture_headers and "requestContext" in event:
+        result["requestContext"] = event["requestContext"]
```

Comment:
This is not standard fare for the `request` piece of the context dictionary. In [the original data structure](https://docs.aws.amazon.com/lambda/latest/dg/services-apigateway.html) it's named `requestContext`. On the one hand, the camelCase makes me uncomfortable and doesn't match our spec at all. On the other hand, it's easy to identify where it comes from, if you're used to dealing with AWS lambdas. Should I rename it or leave it?

Reply:
Good question. I'm tending towards underscores to be in sync with all other fields, but I see your point. Maybe worth bringing it up with the rest of the team, as this probably needs to go through our "adding new fields to context" process sooner or later.